ztoc/ztoc_test.go

```diff
@@ -301,37 +302,45 @@ func TestZtocGeneration(t *testing.T) {
 				testutil.File("filed", string(testutil.RandomByteData(100))),
 				testutil.File("filee", string(testutil.RandomByteData(989993))),
 			},
-			spanSize:  65535,
-			targzName: "testcase0.tar.gz",
-			buildTool: "AWS SOCI CLI",
+			spanSize: 65535,
+			// uncompressedArchiveSize is part of regression testing. The value comes
+			// from running the test before refactoring the way we track this value
+			// to verify that the refactor did not change anything. It doesn't directly
+			// validate that this number is correct.
+			uncompressedArchiveSize: 11975168,
```

Comment:
Can we calculate the uncompressed archive size using a helper (read gzip, uncompress, and get the size). I think it's okay even if we need to write the uncompressed tar to /tmp, since it's just for testing.

This makes the test difficult to understand and add more testcases in the future (e.g. for other compression/tar support).

Reply:
This test already takes quite a long time, so I'd prefer not to add another decompression round or disk roundtrip if we can avoid it.

We can switch from using
```
tarReader := testutil.BuildTarGz(tc.tarEntries, gzip.DefaultCompression)
```

to something like this
```
tarReader := testutil.BuildTar(tc.tarEntries)
countReader := positionTrackerReader{r: tarReader}
gzipReader := testutil.BuildGzip(tarReader, gzip.DefaultCompression)
```

But this is using the exact same counting method as the ztoc itself. Does that actually test what we want?

Reply:
> This test already takes quite a long time, so I'd prefer not to add another decompression round or disk roundtrip if we can avoid it.

Will this add much overhead? I don't have very strong preference but if not, I would prefer calculating from the uncompressed tar file.

> But this is using the exact same counting method as the ztoc itself.

Similarly, since the PR will change the uncompression size calculation to using `positionTrackerReader`, what if we want to add a new testcase later?

Reply:
also just FYI I'm making these `ztoc_test.go` tests generic to different compressions as part of the uncompression layer support work. So far I've made them support both gzip and tar. This also makes it easier to test zstd when we work on that.

(https://github.com/awslabs/soci-snapshotter/commit/c0c6f68d18a50cdd08f4bb6e489da4566ff624f0#diff-a969359eb5e2603a81addbef098db6684a517aad11bc68c5aeb9e91702a37ae1).

Reply:
What do you think about removing these test changes for now, but keeping https://github.com/awslabs/soci-snapshotter/issues/584 open and adding the checks once your changes land?

Reply:
I'm okay with either keeping these test changes (I can rebase in my change) or removing.

just want to figure out how to add checks/new tests after switching to use `positionTrackerReader` to calculate uncompressed archive size (as in your PR). Since once it lands we can no longer do `The value comes from running the test before refactoring the way we track`.

Reply:
I have to push another version to change the variable name for your other comment, so I'll remove the tests changes. I'm going to keep #584 open and I'll add a new check after your changes.